src/i_winmusic.c

```diff
@@ -1357,23 +1319,96 @@ static void FillBuffer(void)
 
 static DWORD WINAPI PlayerProc(void)
 {
-    HANDLE events[2] = { hBufferReturnEvent, hExitEvent };
-
     while (1)
     {
-        switch (WaitForMultipleObjects(2, events, FALSE, INFINITE))
+        if (WaitForSingleObject(hBufferReturnEvent, INFINITE) != WAIT_OBJECT_0)
         {
-            case WAIT_OBJECT_0:
-                FillBuffer();
+            continue;
+        }
+
+        EnterCriticalSection(&CriticalSection);
+
+        buffer.position = 0;
+
+        switch (win_midi_state)
+        {
+            case STATE_STARTUP:
+                ResetDevice();
+                StreamOut();
+                song.rpg_loop = IsRPGLoop();
+                win_midi_state = STATE_PLAYING;
                 break;
 
-            case WAIT_OBJECT_0 + 1:
+            case STATE_SHUTDOWN:
+                // Send notes/sound off prior to reset to prevent volume spikes.
+                SendNotesSoundOff();
+                ResetDevice();
+                StreamOut();
+                win_midi_state = STATE_EXIT;
+                break;
+
+            case STATE_EXIT:
+                LeaveCriticalSection(&CriticalSection);
```

Comment:
I prefer the current behavior because I don't have to scan a different section of code to see what happens on exit.

Reply:
And I prefer single-entry single-exit principle. RAII would be optimal but that's not an option in C. @fabiangreffrath opinion?

Reply:
Wouldn't we achieve the same that the current code does with a mere `break` here?

Reply:
> Wouldn't we achieve the same that the current code does with a mere `break` here?

No, we're in an infinite loop.

Reply:
Eek, I missed the `while (1)` part indeed. I'd be fine with both `goto exit` and the current solution. My main concern is that the code works as expected and is readable by the people who work with it. I am not hunting programming paradigms.

Reply:
Fixed [here](https://github.com/chocolate-doom/chocolate-doom/compare/5b3e48c578b0c2279215849e564a8e99fea741be..026a296399b882fedfe371a13513b1e9a1662ada).

Reply:
Not quite what I had in mind. The point of the `goto` is to go to the point just before the `LeaveCriticalSection` call. This way there is one enter and exactly one matching leave call.

Reply:
It seems easy to add code later that accidentally calls `LeaveCriticalSection()` without calling `EnterCriticalSection()` first, causing the program to hang.

Reply:
No, the point is that those calls are always matched 1-to-1 and control flow is arranged so that it always goes through them.

I usually have a `keep_going` variable or similar. It's set to true at initialization and when the loop wants to exit it's set to false.

Reply:
> I think in this case a `goto` would be better.

I think I'm going to continue messing this up. Could you provide the _exact_ changes you want?

Reply:
This is the `keep_going` idea, if I understand correctly:
<details>

```diff
diff --git a/src/i_winmusic.c b/src/i_winmusic.c
index ae6ffe09..a2098463 100644
--- a/src/i_winmusic.c
+++ b/src/i_winmusic.c
@@ -1444,7 +1444,9 @@ static void FillBuffer(void)
 
 static DWORD WINAPI PlayerProc(void)
 {
-    while (1)
+    static boolean keep_going = true;
+
+    while (keep_going)
     {
         if (WaitForSingleObject(hBufferReturnEvent, INFINITE) != WAIT_OBJECT_0)
         {
@@ -1475,8 +1477,8 @@ static DWORD WINAPI PlayerProc(void)
                 break;
 
             case STATE_EXIT:
-                LeaveCriticalSection(&CriticalSection);
-                goto exit;
+                keep_going = false;
+                break;
 
             case STATE_PLAYING:
                 if (update_volume)
@@ -1507,7 +1509,6 @@ static DWORD WINAPI PlayerProc(void)
         LeaveCriticalSection(&CriticalSection);
     }
 
-exit:
     return 0;
 }
 
```
</details>

Reply:
That is exactly what I had in mind.

Reply:
Fixed [here](https://github.com/chocolate-doom/chocolate-doom/compare/d2f57cc3db1b3905635dc136d6f0503525f2d715..40e7e23803ef8b7e2ac0c75046771fc9589e0329).